### PR TITLE
Remove Canonical Attestations From Pool Correctly

### LINKED ISSUE #9443: Only delete canonical attestations from pending pool
# Description

Beacon node deletes the attestations in beacon block from the pending pool as soon as the beacon block has passed p2p validation. It'll be better that Beacon node only deletes attestations in the beacon block from the pending pool when the beacon block is marked canonical

### PR DESCRIPTION
**Pull request by terencechain**
Fixes #9443

### CODE REVIEW

File: beacon-chain/blockchain/process_block.go [hunk 1]
```diff
@@ -413,3 +417,29 @@ func (s *Service) savePostStateInfo(ctx context.Context, r [32]byte, b block.Sig
 	}
 	return nil
 }
+
+// This removes the attestations from the mem pool. It will only remove the attestations if input root `r` is canonical,
+// meaning the block `b` is part of the canonical chain.
+func (s *Service) removeBlockAttestationsInPool(ctx context.Context, r [32]byte, b block.SignedBeaconBlock) error {
```
Comment (terencechain) [inline beacon-chain/blockchain/process_block.go:423]: Opted in to save one HTR by having the caller passing in precomputed hash tree root `r`

Another option is to `b.Block().HashTreeRoot()` within the function but not sure about the performance implication during batch block processing so I opted out

**Review by prestonvanloon — CHANGES_REQUESTED** [review]
Please add feature flag for testing

**Review by prestonvanloon — APPROVED** [review]

File: beacon-chain/blockchain/process_block.go [hunk 2]
```diff
@@ -140,6 +140,12 @@ func (s *Service) onBlock(ctx context.Context, signed block.SignedBeaconBlock, b
 			log.WithError(err).Warn("Could not update head")
 		}
 
+		if featureconfig.Get().CorrectlyPruneCanonicalAtts {
```
Comment (rauljordan) [inline beacon-chain/blockchain/process_block.go:143]: Let's move this condtional inside the function removeBlockAttstationsInPool

File: beacon-chain/blockchain/process_block.go [hunk 3]
```diff
@@ -413,3 +419,29 @@ func (s *Service) savePostStateInfo(ctx context.Context, r [32]byte, b block.Sig
 	}
 	return nil
 }
+
+// This removes the attestations from the mem pool. It will only remove the attestations if input root `r` is canonical,
+// meaning the block `b` is part of the canonical chain.
+func (s *Service) removeBlockAttestationsInPool(ctx context.Context, r [32]byte, b block.SignedBeaconBlock) error {
```
Comment (rauljordan) [inline beacon-chain/blockchain/process_block.go:425]: Can we rename this to pruneCanonicalAttestationsFromPool?

File: beacon-chain/sync/subscriber_beacon_blocks.go [hunk 4]
```diff
@@ -37,12 +38,13 @@ func (s *Service) beaconBlockSubscriber(ctx context.Context, msg proto.Message)
 		return err
 	}
 
-	// Delete attestations from the block in the pool to avoid inclusion in future block.
-	if err := s.deleteAttsInPool(block.Body().Attestations()); err != nil {
-		log.Debugf("Could not delete attestations in pool: %v", err)
-		return nil
+	if !featureconfig.Get().CorrectlyPruneCanonicalAtts {
```
Comment (rauljordan) [inline beacon-chain/sync/subscriber_beacon_blocks.go:41]: Same feedback, let's move this conditional inside the function

Comment (terencechain) [reply beacon-chain/sync/subscriber_beacon_blocks.go:41]: This is a different function

Comment (rauljordan) [reply beacon-chain/sync/subscriber_beacon_blocks.go:41]: can we move the correctlyprunecanonicalatts conditional inside of deleteAttsInPool ?

**Review by rauljordan — APPROVED** [review]

**Pull request merged**